pos_sale_order/models/pos_session.py

```diff
@@ -83,29 +81,43 @@ def _create_bank_statement_line(self, statement, payments):
                 _("Following sales are not invoiced, please invoice it: \n- %s")
                 % ("\n- ".join(sale_missing_invoice.mapped("name")))
             )
-        move_lines = (
-            bk_line.move_id.line_ids
-            + payments.pos_sale_order_id.invoice_ids.filtered(
-                lambda s: s.state == "posted"
-            ).line_ids
+        return payments.pos_sale_order_id.invoice_ids.filtered(
+            # skip canceled invoice
+            lambda s: s.state
+            == "posted"
+        ).line_ids.filtered(
+            # get receivable line
+            lambda s: s.account_id
+            == payments.payment_method_id.receivable_account_id
         )
-        return move_lines.filtered(
+
+    def _create_bank_statement_line(self, statement, payments):
+        vals = self._prepare_sale_statement(payments)
+        vals["statement_id"] = statement.id
+        bk_line = self.env["account.bank.statement.line"].create(vals)
+        return bk_line.line_ids.filtered(
             lambda s: s.account_id == payments.payment_method_id.receivable_account_id
         )
 
     def _create_bank_statement_line_and_reconcile(self):
-        to_reconcile = []
         self.ensure_one()
         payments = self.env["pos.payment"].search([("session_id", "=", self.id)])
         grouped_payments = defaultdict(
             lambda: defaultdict(lambda: self.env["pos.payment"])
         )
+        # Depending of the configuration we split the payment per sale order
+        # or by invoice
+        # Grouping by invoice make it easier to reconcile and to understand what
+        # have been reconciled with
         for payment in payments:
             if payment.payment_method_id.split_transactions:
                 key = payment.pos_sale_order_id
             else:
-                key = payment.partner_id
+                key = payment.pos_sale_order_id.invoice_ids
```

Comment:
TODO
We can improve this by adding in the ref of the move line of the payment ths invoice or sale order ref

Reply:
https://github.com/akretion/pos-sale-order/pull/52 ?

Reply:
no, what I was meaning, is that Odoo do not set anything in the field "ref" of the payment move.
So if the accountant at some point unreconciled a payment done by the POS it's super hard to reconcile it.
To improve the user experience, now I put
- the sale order name  + the invoice name (in case of transaction split by sale order)
- the invoice none split transaction, ex: cash)